code/datums/exonet.dm

```diff
@@ -0,0 +1,226 @@
+			if(my_client)
+				my_client.stored_login = EA.login
+				my_client.stored_password = EA.password
+
+/datum/exonet/proc/find_email_by_name(var/login)
+	for(var/datum/computer_file/data/email_account/A in email_accounts)
+		if(A.login == login)
+			return A
+	return 0
+
+/datum/exonet/proc/is_connected_plexus()
+	for(var/obj/machinery/computer/exonet/uplink/modem in modems)
+		if(modem.operable())
+			return TRUE
+	return FALSE
+
+/datum/exonet/proc/add_device(var/device, var/keydata)
+	if(!router)
+		return 0 // Uh?? No router? Guess the network is busted.
+	if(router.keydata != keydata)
+		return 0 // Authentication failed.
+
+	if(istype(device, /obj/machinery/computer/exonet/mainframe))
+		mainframes |= device
+	else if(istype(device, /obj/machinery/computer/exonet/broadcaster))
+		broadcasters |= device
+	else if(istype(device, /obj/machinery/computer/exonet/uplink))
+		modems |= device
+	else if(istype(device, /obj/machinery/computer/exonet/broadcaster/router) && !router)
+		router = device // Special setty-uppy-timy for routers.
+	network_devices |= device
+
+/datum/exonet/proc/remove_device(var/device)
+	if(istype(device, /obj/machinery/computer/exonet/mainframe))
+		mainframes -= device
+	else if(istype(device, /obj/machinery/computer/exonet/broadcaster))
+		broadcasters -= device
+	else if(istype(device, /obj/machinery/computer/exonet/uplink))
+		modems -= device
+	network_devices -= device
+
+/datum/exonet/proc/set_router(var/device)
+	if(router)
+		// THERE CAN BE ONLY ONE!!!
+		LAZYREMOVE(network_devices, router)
+		LAZYREMOVE(broadcasters, router)
+	router = device
+	LAZYADD(network_devices, device)
+	LAZYADD(broadcasters, device)
+
+/datum/exonet/proc/check_banned(var/nid)
+	return nid in banned_nids
+
+// Simplified logging: Adds a log. log_string is mandatory parameter, source is optional.
+/datum/exonet/proc/add_log(var/log_string, var/obj/item/stock_parts/computer/network_card/source = null)
+	if(!log_server)
+		return // No log server? No logs.
+	var/datum/computer_file/data/logfile/logfile = log_server.get_log_file()
+	if(!istype(logfile))
+		return // No log file, or no space to create one. Or some other error.
+
+	var/log_text = "[stationtime2text()] - "
+	if(source)
+		log_text += "[source.get_network_tag()] - "
+	else
+		log_text += "*SYSTEM* - "
+	log_text += log_string
+	var/list/logs = splittext(logfile.stored_data, "\[br\]")
+	if(length(logs) >= log_server.setting_max_log_count)
+		logs.Cut(1, 2)
+	logs.Add(log_text)
+	logfile.stored_data = jointext(logs, "\[br\]")
+
+/datum/exonet/proc/get_signal_strength(var/obj/device, var/netspeed)
+	var/best_signal = -1
+	var/turf/device_turf = get_turf(device)
+	if(!device_turf)
+		return best_signal
+	for(var/obj/machinery/computer/exonet/broadcaster/broadcaster in broadcasters)
+		var/turf/b_turf = get_turf(broadcaster)
+		if(b_turf.z != device_turf.z || !broadcaster.operable())
```

Comment:
are broadcasters intended to be one per z-level, rather than one per connected-zlevels?

Reply:
Yes.

Reply:
the link between broadcasters is wuju space transmitters, able to link to each other across vast distances in space. That tech is bulky. The wifi transmitters fittable in portable devices cannot do that reasonably, or whatever. And thus their range is smaller.

You could put a relay on an exoplanet and connect it to Torch's network, for example.